File: python/15-link-list-3.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class LinkList:
    def __init__(self):
        self.head_node = None
        self.length = 0

    def insert(self,data):
        new_node = Node(data)
        self.length += 1
        if self.head_node is None:
            self.head_node = new_node
            return self.head_node
        
        current_node = self.head_node
        while current_node.next is not None:
            current_node =current_node.next
        current_node.next = new_node
        return new_node

    def print_link_list(self):
        current_node = self.head_node
        while current_node is not None:
            if current_node.next is not None:
                print(current_node.data,end=" ")
            else:
                print(current_node.data)
            current_node = current_node.next

    def get(self,position):
        
        if position < 1 or position > self.length:
            return None
        i = 1
        current_node = self.head_node
        while current_node is not None:

            if i == position:
                return current_node
            i += 1
            current_node = current_node.next
        return None
    
    
    def insert_at(self,n,data):
        new_node = Node(data)
        if n==1:
            new_node.next = self.head_node
            self.head_node = new_node
            self.length += 1
            return new_node
        else:
            preNode = self.get(n-1)
            if preNode is not None:
                new_node.next = preNode.next
                preNode.next = new_node
                self.length += 1
                return new_node
        return None

    def delete(self,n):
        if self.head_node is None:
            return None
        if n == 1:
            delete_node = self.head_node
            self.head_node = self.head_node.next
            self.length -= 1
            return delete_node
        else:
            pre_node = self.get(n-1)
            if pre_node is not None and pre_node.next is not None:
                delete_node = pre_node.next
                pre_node.next = pre_node.next.next
                self.length -= 1
                return delete_node
        return None


k = int(input())
nums = list(map(int, input().split()))
link_list = LinkList()

for num in nums:
    link_list.insert(num)
            
s = int(input())

for _ in range(s):

    insert_data = list(map(int, input().split()))

        
    objective_node = link_list.insert_at(insert_data[0], insert_data[1])

    if objective_node is not None:
            
        link_list.print_link_list()
    else:
        print("Insertion position is invalid.")
    
l = int(input())

for _ in range(l):

    delete_position = int(input())
        
    deleted_node = link_list.delete(delete_position)
    if deleted_node is not None:
        link_list.print_link_list()
    else:
        print("Deletion position is invalid.")

File: Java/Java-base/15-LinkList3.java
import java.util.Scanner;

class LinkList {
    public class Node {
        int data;
        Node next;
        public Node(int data) {
            this.data = data;
            this.next = null;
        }
    }
    private Node headNode;
    private int length;
    public LinkList() {
        this.headNode = null;

    }

    public Node insert(int data ) {
        this.length++;
        Node newNode = new Node(data);
        if (headNode == null) {
            headNode = newNode;
            return headNode;
        }
        else {
            Node currentNode = headNode;
            while (currentNode.next !=null) {
                currentNode = currentNode.next;

            }
            currentNode = newNode;
        }
        return null;
    }

    public void printLinkList() {
        Node currentNode = headNode;

        while (currentNode != null) {
            System.out.printf("%d%c",currentNode,currentNode.next == null ? '\n' : ' ');
            currentNode = currentNode.next;
        }
    }

    public Node get(int n) {
        if (n<1||n<length) {
            return null;
        }
        else {
            Node currentNode = headNode;
            int i = 1;
            while (currentNode != null) {
                if(i==n) {
                    return currentNode;
                }
                i++;            
            }
        }
        return null;
    }

    public Node add(int n,int data) {
        this.length++;
        Node node = new Node(data);
        if (headNode == null) {
            headNode = node;
            return headNode;
        }
        else {
            Node preNode = get(n-1);
            if (preNode != null && preNode.next !=null) {
                node.next = preNode.next;
                preNode.next =node;
                return node; 
            }
            else {
                return null;
            }
        }
        return null;
    }

    public Node deleteNode(int n) {
         if (headNode !=null) {
            return null;
         }
         else {
            Node preNode = get(n-1);
            Node deleteNode = get(n);
            if (preNode != null && deleteNode !=null) {
                preNode.next = preNode.next.next;
                this.length--;
                return preNode.next;
            }
            else {
                return null;
            }
         }
        return null;
            
    }
}
public class Main {
    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);
        LinkList linklist = new LinkList();
                while (sc.hasNextInt()) {
            int k = sc.nextInt();
            for (int i= 0;i<k;i++) {
                linklist.insert(sc.nextInt()) ;
            }

            int S = sc.nextInt();
            int n,x = 0;
            for(int j=0 ;j<S;j++) {
                n= sc.nextInt();
                x= sc.nextInt();
                LinkList.Node node =linklist.add(n,x);
                if(node != null) {
                    linklist.printLinkList();
                }
                else {
                    System.out.println("Insertion position is invalid.");
                }

            }

            int L = sc.nextInt();
            int m = 0;
            
            for (int h=0;h<L;h++) {
                m = sc.nextInt();
                LinkList.Node deleteNode =linklist.delete(m);
                if(deleteNode != null) {
                    linklist.printLinkList();
                }
                else {
                    System.out.println("Deletion position is invalid.");
                }
            }
        }
        sc.close();
    }
}
